lib/auditd/auditd_linux.go

```diff
@@ -0,0 +1,288 @@
+	if err != nil {
+		return 0, trace.ConvertSystemError(err)
+	}
+
+	loginuid, err := strconv.ParseInt(string(data), 10, 64)
+	if err != nil {
+		return 0, trace.Wrap(err)
+	}
+	return loginuid, nil
+}
+
+// SendEvent sends a single auditd event. Each request create a new netlink connection.
+// This function returns no error if it runs with no root permissions.
+func SendEvent(event EventType, result ResultType, msg Message) error {
+	if !hasCapabilities() {
+		// Disable auditd when not running as a root.
+		return nil
+	}
+
+	msg.SetDefaults()
+
+	client := NewClient(msg)
+	defer func() {
+		err := client.Close()
+		if err != nil {
+			log.WithError(err).Error("failed to close auditd client")
+		}
+	}()
+
+	if err := client.SendMsg(event, result); err != nil {
+		return trace.Wrap(err)
+	}
+
+	return nil
+}
+
+func (c *Client) connect() error {
+	c.mtx.Lock()
+	defer c.mtx.Unlock()
+
+	if !c.enabled && c.conn != nil {
+		return ErrAuditdDisabled
+	}
+
+	conn, err := c.dial(syscall.NETLINK_AUDIT, nil)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	c.conn = conn
+
+	status, err := getAuditStatus(c.conn)
+	if err != nil {
+		return trace.Errorf("failed to get auditd status: %v", trace.ConvertSystemError(err))
+	}
+
+	c.enabled = status.Enabled == 1
+
+	if status.Enabled != 1 {
+		return ErrAuditdDisabled
+	}
+
+	return nil
+}
+
+// NewClient creates a new auditd client. Client is not connected when is returned.
+func NewClient(msg Message) *Client {
+	msg.SetDefaults()
+
+	execName, err := os.Executable()
+	if err != nil {
+		log.WithError(err).Warn("failed to get executable name")
+		execName = "?"
+	}
+
+	// Teleport never tries to get the hostname name.
+	// Let's mimic the sshd behavior.
+	const hostname = "?"
+
+	return &Client{
+		execName:     execName,
+		hostname:     hostname,
+		systemUser:   msg.SystemUser,
+		teleportUser: msg.TeleportUser,
+		address:      msg.ConnAddress,
+		ttyName:      msg.TTYName,
+
+		dial: func(family int, config *netlink.Config) (NetlinkConnecter, error) {
+			return netlink.Dial(family, config)
+		},
+		mtx: &sync.Mutex{},
+	}
+}
+
+func getAuditStatus(conn NetlinkConnecter) (*auditStatus, error) {
+	_, err := conn.Execute(netlink.Message{
+		Header: netlink.Header{
+			Type:  netlink.HeaderType(AuditGet),
+			Flags: netlink.Request | netlink.Acknowledge,
+		},
+		Data: nil,
+	})
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	msgs, err := conn.Receive()
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	if len(msgs) != 1 {
+		return nil, trace.Errorf("returned wrong messages number, expected 1, got: %d", len(msgs))
+	}
+
+	byteOrder := nlenc.NativeEndian()
+	status := &auditStatus{}
+
+	payload := bytes.NewReader(msgs[0].Data[:])
+	if err := binary.Read(payload, byteOrder, status); err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	return status, nil
+}
+
+// SendMsg sends a message. Client will create a new connection if not connected already.
```

Comment:
As Nic mentioned, it currently looks like `SendMsg` and `SendEvent` are always creating a new client rather than using the cached client.

I might be missing something, but I'd recommend changing the API to function like this:

```
client, err := auditd.NewClient()
if err != nil {
  return trace.Wrap(err)
}
defer client.Close()

if err := client.SendEvent(event); err != nil {
  return trace.Wrap(err)
}
```

Then in `auditd.go` you can just introduce a mock client so that creating the client and sending events both produce noops.

Reply:
I was thinking about it, and the initial implementation looked like that. My main issue with this approach is that it is a bit verbose compared to only one function call. Also, you need to remember to close the connection at the end. That's why I decided to wrap everything in a function.

There are, of course, a few more things to consider if we use only `SendEvent()`:
1. If we create a new connection for each message, we probably create more connections when needed. This is not exactly true. In my implementation, I found only one place where we could re-use the connection. I don't think that is a big issue then.
2. Why then my client internally holds a mutex if the connection is not being used from more than one goroutine? Mainly for correctness. the `auditd` package exposes the client, so if anyone decides to use it in the future, it may create a race condition without knowing it.

I think there are three possible actions here:
1. Remove the internal mutex to be explicit about the fact that it is not really needed. Make `NewClient()` constructor private, so the only API in that package is the `SendEvent()` function. That would make the API more explicit. 
2. Leave it as it is. Currently, the `auditd` package exposes two APIs. Higher lever `SendEvent()` function that allows you to send one message and `NewClient()` that can create a new client that is thread-safe.
3. I could use the auditd client in that one place where the connection can be reused to reduce the overhead where possible and make use of the `NewClient()` API.

WDYGT?
@Joerger @nklaassen

Reply:
Looks like the connection is being reused properly now, so looks good to me as is.

> I could use the auditd client in that one place where the connection can be reused to reduce the overhead where possible and make use of the NewClient() API.

I think it would make sense to use `NewClient` in that one place, otherwise we might as well remove the `NewClient` API and simplify the connection logic for a single use connection.